Mathlib/ModelTheory/Equivalence.lean

```diff
@@ -257,6 +273,61 @@ theorem inf_iff_not_sup_not : (φ ⊓ ψ) ⇔[T] (φ.not ⊔ ψ.not).not :=
 
 end Formula
 
+namespace Theory
+
+variable (T) (α)
+
+/-- The type of formulas modulo a theory. -/
+protected def Formula := Quot (T.Iff : L.Formula α → _)
+
+variable {T} {α}
+
+namespace Formula
+
+instance : Coe (L.Formula α) (T.Formula α) := ⟨Quot.mk _⟩
+
+@[simps]
+instance : BooleanAlgebra (T.Formula α) where
+  le := Quot.lift₂ T.Imp
+    (fun _ _ _ h => iff_eq_eq.mp ⟨fun h' => h'.trans h.mp, fun h' => h'.trans h.mpr⟩)
+    (fun _ _ _ h => iff_eq_eq.mp ⟨h.mpr.trans, h.mp.trans⟩)
+  le_refl := Quot.ind Imp.refl
+  le_trans := Quot.ind (fun _ => Quot.ind (fun _ => Quot.ind (fun _ => Imp.trans)))
+  le_antisymm := Quot.ind (fun _ => Quot.ind (fun _ h₁ h₂ =>
+    (T.iffSetoid.iseqv.quot_mk_eq_iff _ _).2 (imp_antisymm h₁ h₂)))
+  sup := Quot.map₂ (· ⊔ ·) (fun _ _ _ => (Iff.refl _).sup)
+     (fun _ _ _ h => h.sup (refl _))
+  le_sup_left := Quot.ind (fun _ => Quot.ind (fun _ => imp_sup_left _ _))
+  le_sup_right := Quot.ind (fun _ => Quot.ind (fun _ => imp_sup_right _ _))
+  sup_le := Quot.ind (fun _ => Quot.ind (fun _ => (Quot.ind (fun _ => sup_imp))))
+  inf := Quot.map₂ (· ⊓ ·) (fun _ _ _ => (Iff.refl _).inf)
+     (fun _ _ _ h => h.inf (refl _))
+  inf_le_left := Quot.ind (fun _ => Quot.ind (fun _ => inf_imp_left _ _))
+  inf_le_right := Quot.ind (fun _ => Quot.ind (fun _ => inf_imp_right _ _))
+  le_inf := Quot.ind (fun _ => Quot.ind (fun _ => (Quot.ind (fun _ => imp_inf))))
+  le_sup_inf := Quot.ind (fun _ => Quot.ind fun _ => Quot.ind (fun _ M v xs h => by
+    simp only [BoundedFormula.realize_inf, BoundedFormula.realize_sup] at *
+    tauto))
+  bot := Quot.mk _ ⊥
+  bot_le := Quot.ind bot_imp
+  top := Quot.mk _ ⊤
+  le_top := Quot.ind imp_top
+  compl := Quot.map Formula.not (fun _ _ h => h.not)
+  inf_compl_le_bot := Quot.ind (fun φ => φ.inf_not_iff_bot.mp)
+  top_le_sup_compl := Quot.ind (fun φ => φ.sup_not_iff_top.mpr)
+
+/-- Evaluates a formula modulo `T` as true or false in a nonempty model of `T`. -/
+@[simp]
+def Realize (φ : T.Formula α) {M : Type*} [L.Structure M] [M ⊨ T] [Nonempty M] (v : α → M) : Prop :=
+  Quot.lift (fun (ψ : L.Formula α) => ψ.Realize v) (fun _ _ h => iff_eq_eq.mp h.realize_iff) φ
+
+lemma Realize_coe (φ : L.Formula α) {M : Type*} [L.Structure M] [M ⊨ T] [Nonempty M] (v : α → M) :
```

Comment:
```suggestion
@[simp, norm_cast]
lemma realize_coe (φ : L.Formula α) {M : Type*} [L.Structure M] [M ⊨ T] [Nonempty M] (v : α → M) :
```
maybe?

Reply:
`norm_cast` gives an error (when I use `Antisymmetrization` and when I don't) along the lines of
```
norm_cast: badly shaped lemma, lhs must contain at least one coe
  Realize (toAntisymmetrization T.Imp φ) v
 ```